### PR TITLE
[PyTorch] validate that SparseTensorImpl::dim needn't be overridden

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50176 [PyTorch] Devirtualize TensorImpl::sizes() with macro
* #49770 [PyTorch] Devirtualize TensorImpl::dim() with macro
* **#49767 [PyTorch] validate that SparseTensorImpl::dim needn't be overridden**
* #49766 [PyTorch] Devirtualize TensorImpl::numel() with macro

I'm told that the base implementation should work fine. Let's validate that in an intermediate diff before removing it.

Differential Revision: [D25686830](https://our.internmc.facebook.com/intern/diff/D25686830/)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 1]
```diff
@@ -70,6 +70,7 @@ void SparseTensorImpl::set_storage_offset(int64_t storage_offset) {
 }
 
 int64_t SparseTensorImpl::dim() const {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(sparse_dim_ + dense_dim_ == TensorImpl::dim());
```
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp:73]: why debug only?  Do we have full coverage for our debug builds?

Comment (swolchok) [reply aten/src/ATen/SparseTensorImpl.cpp:73]: beats me, but slowing down `dim()` in prod seems like a bad choice.

**Review by bhosmer — APPROVED** [review]

File: aten/src/ATen/SparseTensorImpl.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/SparseTensorImpl.cpp:73]: right, but you don't have to land it.  This would just ensure our OSS CI at least runs over it.

Comment (swolchok) [reply aten/src/ATen/SparseTensorImpl.cpp:73]: done https://github.com/pytorch/pytorch/pull/50172

**Comment by swolchok** [comment]
https://github.com/pytorch/pytorch/pull/50172 looks fine. a single test failed due to taking too long

**Pull request closed without merging**